### PR TITLE
Add transformation to have disjuncts fixed to their current values.

### PR DESCRIPTION
**Pull request by qtothec**
Also small change to help with error tracing.

### CODE REVIEW

**Comment by jsiirola** [comment]
I think there are some logical issues with this transformation.  In particular, I would maintain that if you call `.apply_to(model.disjunct[i])` the wrong thing will happen: the individual disjunct will be transformed (without error), but not the other Disjuncts in the Disjunction.  Similarly, because you walk the model looking for Disjuncts, you can get strange behavior when the model has Disjuncts that do not appear in any Disjunction.  I think I would rather see this transformation walk the model looking for Disjunctions, and only transform the Disjuncts that are used in the identified (active) Disjunctions.

**Comment by qtothec** [comment]
Right. Wasn't my use case, so I hadn't considered it, but I'll make those changes.

**Comment by qtothec** [comment]
@jsiirola Updated disjunct fixing code

**Comment by qtothec** [comment]
@jsiirola Any other changes you want on this?

File: pyomo/gdp/plugins/fix_disjuncts.py [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+from pyomo.core.base.constraint import Constraint
+from pyomo.core.kernel.numvalue import value
+from pyomo.core.kernel.component_set import ComponentSet
+from pyomo.gdp import GDP_Error
+from pyomo.gdp.disjunct import (Disjunct, Disjunction, _DisjunctData,
+                                _DisjunctionData)
+from pyomo.util.plugin import alias
+from six import itervalues
+
+__author__ = "Qi Chen <https://github.com/qtothec>"
```
Comment (jsiirola) [inline pyomo/gdp/plugins/fix_disjuncts.py]: Remove, please.

File: pyomo/gdp/plugins/fix_disjuncts.py [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+logger = logging.getLogger('pyomo.core')
```
Comment (jsiirola) [inline pyomo/gdp/plugins/fix_disjuncts.py]: Log to `pyomo.gdp.fix_disjuncts`

File: pyomo/gdp/plugins/fix_disjuncts.py [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+
+class GDP_Disjunct_Fixer(Transformation):
+    """Fix disjuncts to Blocks."""
```
Comment (jsiirola) [inline pyomo/gdp/plugins/fix_disjuncts.py]: This is not terribly descriptive.

File: pyomo/gdp/plugins/fix_disjuncts.py [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+            obj.indicator_var.fix(1)
+            self._transformContainer(obj)
+        elif fabs(value(obj.indicator_var)) <= self.integer_tolerance:
+            obj.parent_block().reclassify_component_type(obj, Block)
+            obj.indicator_var.fix(0)
+            # Deactivate all constituent constraints and disjunctions
+            for constr in obj.component_objects(
+                    ctype=(Constraint, Disjunction),
+                    active=True, descend_into=True):
+                constr.deactivate()
```
Comment (jsiirola) [inline pyomo/gdp/plugins/fix_disjuncts.py]: Why not just deactivate the entire block?

**Review by jsiirola — CHANGES_REQUESTED** [review]
In addition to the comments, can you provide some tests?

File: pyomo/gdp/plugins/fix_disjuncts.py (hunk 4 above)
Comment (qtothec) [reply pyomo/gdp/plugins/fix_disjuncts.py]: I'm paranoid about the solvers not detecting variables on deactivated blocks.

Comment (jsiirola) [reply pyomo/gdp/plugins/fix_disjuncts.py]: Ack.  You are absolutely correct.  There is a flipside, though: Disjuncts with ActiveComponents other than Constraint that aren't deactivated when you turn the disjunct off.  We really need to file a ticket and force that fix for Pyomo 5.5.  Can you at least document here that this is a HACK to get around the deactivated blocks stupidity in the writers?

Comment (qtothec) [reply pyomo/gdp/plugins/fix_disjuncts.py]: Sure. It also depends on the outcome of the discussion in https://github.com/Pyomo/pyomo/issues/309.

**Review by jsiirola — APPROVED** [review]

**Pull request merged**